perfkitbenchmarker/linux_benchmarks/cloudsuite_web_serving_benchmark.py

```diff
@@ -316,68 +71,108 @@ def Prepare(benchmark_spec):
-    latency99 = re.findall(r'\<p99th\>(\d+\.\d*)', stdout)
-    return sum_ops_per_sec, latency, latency99
+
+  def PrepareCommon(vm):
+    if not docker.IsInstalled(vm):
+      vm.Install('docker')
+    vm.RemoteCommand("sudo sh -c 'echo %s web_server >>/etc/hosts'" %
+                     frontend.internal_ip)
+    vm.RemoteCommand("sudo sh -c 'echo %s memcache_server >>/etc/hosts'" %
+                     frontend.internal_ip)
+    vm.RemoteCommand("sudo sh -c 'echo %s mysql_server >>/etc/hosts'" %
+                     backend.internal_ip)
+    vm.RemoteCommand("sudo sh -c 'echo %s faban_client >>/etc/hosts'" %
+                     client.internal_ip)
+
+  def PrepareFrontend(vm):
+    PrepareCommon(vm)
+    vm.RemoteCommand('sudo docker pull cloudsuite/web-serving:web_server')
+    vm.RemoteCommand('sudo docker pull cloudsuite/web-serving:memcached_server')
+    vm.RemoteCommand('sudo docker run -dt --net host --name web_server '
+                     'cloudsuite/web-serving:web_server /etc/bootstrap.sh')
+    vm.RemoteCommand('sudo docker run -dt --net host --name memcache_server '
+                     'cloudsuite/web-serving:memcached_server')
+
+  def PrepareBackend(vm):
+    PrepareCommon(vm)
+    vm.RemoteCommand('sudo docker pull cloudsuite/web-serving:db_server')
+    vm.RemoteCommand('sudo docker run -dt --net host --name mysql_server '
+                     'cloudsuite/web-serving:db_server')
+
+  def PrepareClient(vm):
+    PrepareCommon(vm)
+    vm.RemoteCommand('sudo docker pull cloudsuite/web-serving:faban_client')
+
+  target_arg_tuples = [(PrepareFrontend, [frontend], {}),
+                       (PrepareBackend, [backend], {}),
+                       (PrepareClient, [client], {})]
+  vm_util.RunParallelThreads(target_arg_tuples, len(target_arg_tuples))
 
 
 def Run(benchmark_spec):
+  """Run the web serving benchmark.
+
+  Args:
+    benchmark_spec: The benchmark specification. Contains all data that is
+        required to run the benchmark.
+
+  Returns:
+    A list of sample.Sample objects.
+  """
+  frontend = benchmark_spec.vm_groups['frontend'][0]
   client = benchmark_spec.vm_groups['client'][0]
-  set_faban_home = ('export FABAN_HOME=%s && cd %s && ./run.sh')
-  client.RobustRemoteCommand(set_faban_home % (FABAN_HOME, FABAN_HOME))
   results = []
-  sum_ops_per_sec, latency, latency99 = ParseOutput(client)
-  results.append(sample.Sample('Operations per second',
-                 sum_ops_per_sec, 'ops/s'))
-  results.append(sample.Sample('90th percentile latency HomePage ',
-                 float(latency[0]), 'seconds'))
-  results.append(sample.Sample('99th percentile latency HomePage',
-                 float(latency99[0]), 'seconds'))
+
+  cmd = ('sudo docker run --net host --name faban_client '
+         'cloudsuite/web-serving:faban_client %s %s' %
+         (frontend.internal_ip, FLAGS.cloudsuite_web_serving_load_scale))
+  stdout, _ = client.RemoteCommand(cmd)
```

Comment:
Can you put in a comment what you expect the output to look like?  It's hard to tell what you're doing here.